src/main/java/wolox/training/security/CustomAuthProvider.java

```diff
@@ -0,0 +1,45 @@
+package wolox.training.security;
+
+import java.util.Collections;
+import java.util.Optional;
+import org.springframework.security.authentication.AuthenticationProvider;
+import org.springframework.security.authentication.BadCredentialsException;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.stereotype.Component;
+import wolox.training.exceptions.UserNotFoundException;
+import wolox.training.models.User;
+import wolox.training.repositories.UserRepository;
+
+@Component
+public class CustomAuthProvider implements AuthenticationProvider {
+
+    private final UserRepository userRepository;
+
+    public CustomAuthProvider(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
+
+    @Override
+    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+        String name = authentication.getName();
+        String password = authentication.getCredentials().toString();
+
+        Optional<User> user = userRepository.findByUsername(name);
+
+        if (user.isEmpty()) {
+            throw new UserNotFoundException();
+        } else if (!user.get().validPassword(password)) {
+            throw new BadCredentialsException("Provide password doesn't match user's password");
```

Comment:
Generally for security purposes, instead of saying "The password doesn't match the user's password", we say "The provided user or password is incorrect", whithout specifying which one :) 
Just as a comment for the future, you don't need to change it right now